ext/page_log/palm/palm_kv.c

```diff
@@ -372,6 +373,110 @@ palm_kv_put_page(PALM_KV_CONTEXT *context, uint64_t table_id, uint64_t page_id,
+  PALM_KV_CONTEXT *context, WT_ITEM *item, uint64_t checkpoint_lsn, uint64_t table_id, size_t *size)
+{
+    MDB_cursor *cursor;
+    MDB_stat stat;
+    int ret;
+    size_t count = 0;
+    cursor = NULL;
+    MDB_val kval;
+    MDB_val vval;
+    PAGE_KEY page_key;
+
+    memset(&kval, 0, sizeof(kval));
+    memset(&vval, 0, sizeof(vval));
+    memset(&page_key, 0, sizeof(page_key));
+    page_key.table_id = table_id;
+    swap_page_key(&page_key, &page_key);
+    kval.mv_size = sizeof(page_key);
+    kval.mv_data = &page_key;
+
+    if ((ret = mdb_cursor_open(context->lmdb_txn, context->env->lmdb_pages_dbi, &cursor)) != 0)
+        return (ret);
+
+    /* Get maximum size for page ids */
+    if ((ret = mdb_stat(context->lmdb_txn, context->env->lmdb_pages_dbi, &stat)) != 0)
+        return (ret);
+
+    /* If no entries found, return an error  */
+    if (stat.ms_entries == 0) {
+        item->size = 0;
+        item->data = NULL;
+        return (WT_NOTFOUND);
+    }
+
+    if (item != NULL) {
+        memset(item, 0, sizeof(WT_ITEM));
+        if ((ret = palm_resize_item(item, stat.ms_entries * sizeof(uint64_t))) != 0)
+            return (ret);
+    }
+
+    if (item->data == NULL)
+        return (ENOMEM);
+
+    if ((ret = mdb_cursor_get(cursor, &kval, &vval, MDB_SET_RANGE)) != 0)
+        return (ret);
+
+    uint64_t prev_page_id = 0;
+    int prev_is_tombstone = 0;
+
+    /*
+     * Iterate through the pages table, looking for pages that has a lsn smaller than the given
+     * checkpoint lsn. Note that the pages are sorted by table_id, page_id, lsn in ascending order,
+     * so we are only interested in the last page of each page_id. If the last page is a tombstone,
+     * meaning we're discarding it, then we skip it. If the last page is a full page, we store the
+     * page ID in the item->data array.
+     */
+    while (ret == 0) {
+        if (kval.mv_size != sizeof(PAGE_KEY))
+            return (EINVAL);
+
+        PAGE_KEY *key = (PAGE_KEY *)kval.mv_data;
+        PAGE_KEY decoded_key;
+        swap_page_key(key, &decoded_key);
+
+        /* Skip pages that are not for the requested table. */
+        if (decoded_key.table_id < table_id)
+            /* Skip pages that are not for the requested table. */
+            goto next;
+
+        /* If we have gone past the table, stop. */
+        else if (decoded_key.table_id > table_id)
+            break;
+
+        /* Skip pages that are newer than the checkpoint. */
+        if (decoded_key.lsn >= checkpoint_lsn)
+            goto next;
+
+        /*  If the previous page was not a tombstone, store the previous page ID. */
+        if (prev_page_id != 0 && decoded_key.page_id != prev_page_id && !prev_is_tombstone) {
+            assert(count < stat.ms_entries);
+            ((uint64_t *)item->data)[count++] = prev_page_id;
+        }
+
+        prev_page_id = decoded_key.page_id;
+        prev_is_tombstone = ((decoded_key.flags & WT_PALM_KV_TOMBSTONE) != 0);
+
+next:
```

Comment:
Style: In general it's best to avoid goto labels. This logic can be achieved in the while loop itself. I would refactor this to put the two "next" conditions together and within that call `mdb_cursor_get(MDB_NEXT)` followed by a `continue`.